### PR TITLE
Lsfb 37419 adapter for Azure

### PR DESCRIPTION
**Pull request by rruiz-medtrainer**
# What does this PR?

This PR will provide all the methods to be integrate as an Adapter to league/flysystem-bundle or https://github.com/thephpleague/flysystem, both libraries are compatible with Symfony and Laravel framework

## How to implement this Adapter?

After you install  league/flysystem-bundle on Symfony you can configurate the next way:

```
config/packages/flysystem.yml
# Read the documentation at https://github.com/thephpleague/flysystem-bundle/blob/master/docs/1-getting-started.md
flysystem:
    storages:
        default.storage:
            adapter: 'MedTrainer\Flysystem\AzureBlobStorage\AzureBlobStorageAdapter'
```
and

```
service.yml
MedTrainer\Flysystem\AzureBlobStorage\AzureBlobStorageAdapter:
        class: MedTrainer\Flysystem\AzureBlobStorage\AzureBlobStorageAdapter
        arguments:
            - '@MicrosoftAzure\Storage\Blob\BlobRestProxy'
            - '@logger'
```

### CODE REVIEW

**Review by jledezma-mt — APPROVED** [review]

File: phpunit.xml.dist [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  </testsuites>
+
+  <filter>
+    <whitelist>
+      <directory suffix=".php">src/</directory>
+    </whitelist>
+  </filter>
+
+  <php>
+    <env name="APP_ENV" value="testing"/>
```
Comment (gortega-medtrainer) [inline phpunit.xml.dist:26]: Shouldn't this be test?

**Review by isuarez30 — APPROVED** [review]

**Review by isuarez30 — APPROVED** [review]

**Review by isuarez30 — APPROVED** [review]

File: tests/AzureBlobStorageAdapterTest.php [hunk 2]
```diff
@@ -0,0 +1,480 @@
...
+    ////                        ],
+    ////                    ],
+    ////                    "NextMarker" => null,
+    ////                ]
+    ////            ]));
+    ////        $blobClient = BlobRestProxy::createBlobService('AccountName=devstoreaccount1;AccountKey=Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==;DefaultEndpointsProtocol=http;BlobEndpoint=http://azurite:10000/devstoreaccount1;QueueEndpoint=http://azurite:10001/devstoreaccount1;TableEndpoint=http://azurite:10002/devstoreaccount1;');
+    //        $service = new AzureBlobStorageAdapter($blobClient, $this->logger, 'default');
+    //        $response = $service->listContents('', true);
+    //        $this->assertIsIterable($response, 'Error getting the list of the content');
+    //    }
```
Comment (jledezma-mt) [inline tests/AzureBlobStorageAdapterTest.php:417]: Is this really needed?

File: tests/AzureBlobStorageAdapterTest.php [hunk 3]
```diff
@@ -0,0 +1,480 @@
...
+                ->willThrowException(new UnableToCopyFile(""));
+            $this->expectException(UnableToCopyFile::class);
+        } else {
+            $blobClient->expects(self::any())
+                ->method('copyBlob')
+                ->willReturn(null);
+            $this->expectNotToPerformAssertions();
+        }
+
+        //        $blobClient = BlobRestProxy::createBlobService('AccountName=devstoreaccount1;AccountKey=Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==;DefaultEndpointsProtocol=http;BlobEndpoint=http://azurite:10000/devstoreaccount1;QueueEndpoint=http://azurite:10001/devstoreaccount1;TableEndpoint=http://azurite:10002/devstoreaccount1;');
```
Comment (jledezma-mt) [inline tests/AzureBlobStorageAdapterTest.php:460]: Is this code line needed?

File: src/AzureBlobStorageAdapter.php [hunk 4]
```diff
@@ -1,115 +1,350 @@
...
+        if ($mimetype = $config->get('mimetype')) {
+            $options->setContentType($mimetype);
+        }
+
+        return $options;
+    }
+
+    private function initialize():void
+    {
+        if (!$this->booted) {
```
Comment (gortega-medtrainer) [inline src/AzureBlobStorageAdapter.php:303]: when did we set the booted property? I didn't see it in the createContainer method definition

**Review by jledezma-mt — APPROVED** [review]

File: tests/AzureBlobStorageAdapterTest.php (hunk 3 above)
Comment (rruiz-medtrainer) [reply tests/AzureBlobStorageAdapterTest.php:460]: only about how to test it locally with the service

File: tests/AzureBlobStorageAdapterTest.php (hunk 2 above)
Comment (rruiz-medtrainer) [reply tests/AzureBlobStorageAdapterTest.php:417]: for now, yes, to allow us, continue more easy the test, now we have like 80% of coverage

File: tests/AzureBlobStorageAdapterTest.php [hunk 5]
```diff
@@ -0,0 +1,480 @@
...
+
+    const DIR_TEST = 'mt';
+
+    /** @var MockObject|LoggerInterface  */
+    private $logger;
+
+    public function setUp(): void
+    {
+        $this->logger = $this->createMock(LoggerInterface::class);
+        $this->logger->expects(self::any())
```
Comment (gortega-medtrainer) [inline tests/AzureBlobStorageAdapterTest.php:35]: can we change this to expect the logger to be called at least once?
```suggestion
        $this->logger->expects(self::atLeastOnce())
```

File: tests/AzureBlobStorageAdapterTest.php [hunk 6]
```diff
@@ -0,0 +1,480 @@
...
+        $responseClient->expects(self::any())
+            ->method('getLastModified')
+            ->willReturn(new \DateTimeImmutable());
+        ;
+
+        $blobClient = $this->createMock(BlobRestProxy::class);
+        $blobClient->expects(self::any())
+            ->method('createBlockBlob')
+            ->willReturn($responseClient);
+        //        $blobClient = BlobRestProxy::createBlobService('AccountName=devstoreaccount1;AccountKey=Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==;DefaultEndpointsProtocol=http;BlobEndpoint=http://azurite:10000/devstoreaccount1;QueueEndpoint=http://azurite:10001/devstoreaccount1;TableEndpoint=http://azurite:10002/devstoreaccount1;');
```
Comment (gortega-medtrainer) [inline tests/AzureBlobStorageAdapterTest.php:102]:
```suggestion
```

File: tests/AzureBlobStorageAdapterTest.php [hunk 7]
```diff
@@ -0,0 +1,480 @@
...
+     * @dataProvider getDataWrite
+     */
+    public function testWriteStream($destinationFile, $sourceFile, $config)
+    {
+        $responseClient = $this->createMock(PutBlobResult::class);
+        $responseClient->expects(self::any())
+            ->method('getLastModified')
+            ->willReturn(new \DateTimeImmutable());
+        ;
+        //        $blobClient = BlobRestProxy::createBlobService('AccountName=devstoreaccount1;AccountKey=Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==;DefaultEndpointsProtocol=http;BlobEndpoint=http://azurite:10000/devstoreaccount1;QueueEndpoint=http://azurite:10001/devstoreaccount1;TableEndpoint=http://azurite:10002/devstoreaccount1;');
```
Comment (gortega-medtrainer) [inline tests/AzureBlobStorageAdapterTest.php:125]:
```suggestion
```

File: tests/AzureBlobStorageAdapterTest.php [hunk 8]
```diff
@@ -0,0 +1,480 @@
...
+    {
+        $responseClient = $this->createMock(GetBlobResult::class);
+        $responseClient->expects(self::any())
+            ->method('getContentStream')
+            ->willReturn(fopen($sourceFile, 'r'));
+        $blobClient = $this->createMock(BlobRestProxy::class);
+        $blobClient->expects(self::any())
+            ->method('getBlob')
+            ->willReturn($responseClient);
+        //        $blobClient = BlobRestProxy::createBlobService('AccountName=devstoreaccount1;AccountKey=Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==;DefaultEndpointsProtocol=http;BlobEndpoint=http://azurite:10000/devstoreaccount1;QueueEndpoint=http://azurite:10001/devstoreaccount1;TableEndpoint=http://azurite:10002/devstoreaccount1;');
```
Comment (gortega-medtrainer) [inline tests/AzureBlobStorageAdapterTest.php:173]:
```suggestion
```

File: tests/AzureBlobStorageAdapterTest.php [hunk 9]
```diff
@@ -0,0 +1,480 @@
...
+    {
+        $responseClient = $this->createMock(GetBlobResult::class);
+        $responseClient->expects(self::any())
+            ->method('getContentStream')
+            ->willReturn(fopen($sourceFile, 'r'));
+        $blobClient = $this->createMock(BlobRestProxy::class);
+        $blobClient->expects(self::any())
+            ->method('getBlob')
+            ->willReturn($responseClient);
+        //$blobClient = BlobRestProxy::createBlobService('AccountName=devstoreaccount1;AccountKey=Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==;DefaultEndpointsProtocol=http;BlobEndpoint=http://azurite:10000/devstoreaccount1;QueueEndpoint=http://azurite:10001/devstoreaccount1;TableEndpoint=http://azurite:10002/devstoreaccount1;');
```
Comment (gortega-medtrainer) [inline tests/AzureBlobStorageAdapterTest.php:193]:
```suggestion
```

File: tests/AzureBlobStorageAdapterTest.php [hunk 10]
```diff
@@ -0,0 +1,480 @@
...
+                    'Name' => self::FILE_TEST
+                ]
+            ]
+        ]);
+
+        $blobClient = $this->createMock(BlobRestProxy::class);
+        $blobClient->expects(self::any())
+            ->method('listBlobs')
+            ->willReturn($list);
+        //        $blobClient = BlobRestProxy::createBlobService('AccountName=devstoreaccount1;AccountKey=Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==;DefaultEndpointsProtocol=http;BlobEndpoint=http://azurite:10000/devstoreaccount1;QueueEndpoint=http://azurite:10001/devstoreaccount1;TableEndpoint=http://azurite:10002/devstoreaccount1;');
```
Comment (gortega-medtrainer) [inline tests/AzureBlobStorageAdapterTest.php:258]:
```suggestion
```

File: tests/AzureBlobStorageAdapterTest.php (hunk 2 above)
Comment (gortega-medtrainer) [inline tests/AzureBlobStorageAdapterTest.php:417]:
```suggestion
```

File: tests/AzureBlobStorageAdapterTest.php [hunk 11]
```diff
@@ -0,0 +1,480 @@
...
+                ->willThrowException(new UnableToCopyFile(""));
+            $this->expectException(UnableToCopyFile::class);
+        } else {
+            $blobClient->expects(self::any())
+                ->method('copyBlob')
+                ->willReturn(null);
+            $this->expectNotToPerformAssertions();
+        }
+
+        //        $blobClient = BlobRestProxy::createBlobService('AccountName=devstoreaccount1;AccountKey=Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==;DefaultEndpointsProtocol=http;BlobEndpoint=http://azurite:10000/devstoreaccount1;QueueEndpoint=http://azurite:10001/devstoreaccount1;TableEndpoint=http://azurite:10002/devstoreaccount1;');
```
Comment (gortega-medtrainer) [inline tests/AzureBlobStorageAdapterTest.php:437]:
```suggestion
```

File: tests/AzureBlobStorageAdapterTest.php (hunk 3 above)
Comment (gortega-medtrainer) [inline tests/AzureBlobStorageAdapterTest.php:460]:
```suggestion
```

File: tests/AzureBlobStorageAdapterTest.php [hunk 12]
```diff
@@ -0,0 +1,480 @@
...
+        $this->logger->expects(self::any())
+            ->method('debug');
+        $this->logger->expects(self::any())
+            ->method('error');
+        parent::setUp();
+    }
+
+    public function testCreation(): void
+    {
+        $blobClient = BlobRestProxy::createBlobService('AccountName=devstoreaccount1;AccountKey=Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==;DefaultEndpointsProtocol=http;BlobEndpoint=http://azurite:10000/devstoreaccount1;QueueEndpoint=http://azurite:10001/devstoreaccount1;TableEndpoint=http://azurite:10002/devstoreaccount1;');
```
Comment (gortega-medtrainer) [inline tests/AzureBlobStorageAdapterTest.php:44]: so are we going to actually connect to the service in this tests? will this be temporary? I mean, kind of a ping to the service to assure that we can create the correct instance?

**Review by jrangel736 — APPROVED** [review]

**Pull request merged**